galley/pkg/runtime/processor.go

```diff
@@ -181,33 +152,36 @@ loop:
 		}
 	}
 
-	p.strategy.reset()
+	p.state.close()
 	close(p.stopped)
 	scope.Debugf("Process.process: Exiting process loop")
 }
 
-func (p *Processor) processEvent(e resource.Event) bool {
+func (p *Processor) processEvent(e resource.Event) {
 	scope.Debugf("Incoming source event: %v", e)
-	now := time.Now()
-	recordProcessorEventProcessed(now.Sub(p.lastEventTime))
-	p.lastEventTime = now
-	p.pendingEvents++
+	p.recordEvent()
 
 	if e.Kind == resource.FullSync {
 		scope.Infof("Synchronization is complete, starting distribution.")
-		p.distribute = true
-		return true
+		p.state.onFullSync()
+		return
 	}
 
-	return p.state.apply(e) && p.distribute
+	p.handler.Handle(e)
 }
 
-func (p *Processor) publish() {
+func (p *Processor) recordEvent() {
 	now := time.Now()
-	recordProcessorSnapshotPublished(p.pendingEvents, now.Sub(p.lastSnapshotTime))
-	p.lastSnapshotTime = now
-	sn := p.state.buildSnapshot()
+	recordProcessorEventProcessed(now.Sub(p.lastEventTime))
+	p.lastEventTime = now
+}
 
-	p.distributor.SetSnapshot(snapshot.DefaultGroup, sn)
-	p.pendingEvents = 0
+func buildDispatcher(states ...*State) *processing.Dispatcher {
```

Comment:
Not sure how you're envisioning to use this. Based on the approach you're taking, you'll need exactly one of these for the current path, and a specialized one for the endpoints case, right?

Reply:
My current plan is to have 2 states. The dispatcher will route events to the correct one based on their schema.

Reply:
How are we going to drive this? Additional metadata, or are these going to be hard-wired, specialized implementations?

Reply:
Hard-wired for now until your dataflow code lands.

Reply:
Then you will need a specialized implementation somewhere, for different handling/projection, right? I am trying to understand how routing to the right handler function is going to hapen.